aten/src/ATen/nnapi/nnapi_model_loader.cpp

```diff
@@ -0,0 +1,264 @@
+#include <stdint.h>
+
+#include "NeuralNetworks.h"
+#include "nnapi_wrapper.h"
+#include "nnapi_model_loader.h"
+
+
+#ifndef NNAPI_LOADER_STANDALONE
+
+# include <c10/util/Logging.h>
+
+#else
+
+#define CAFFE_ENFORCE(cond, ...) do { if (!cond) { return -1; } } while (0)
+
+#endif
+
+
+#define NNAPI_CHECK(res) CAFFE_ENFORCE(res == ANEURALNETWORKS_NO_ERROR)
+
+
+namespace caffe2 {
+namespace nnapi {
+
+namespace {
+
+/*
+Serialized format for NNAPI models.  It is basically just a list arguments
+for calls to be made to NNAPI.
+*/
+
+typedef enum _SourceType {
+  SOURCE_IMMEDIATE = 0,
+  SOURCE_NUMBERED_BUFFER = 2,
+  SOURCE_NUMBERED_MEMORY = 3,
+} SourceType;
+
+typedef struct _SerializedOperand {
+  int32_t type;
+  uint32_t dimension_count;
+  float scale;
+  int32_t zero_point;
+} SerializedOperand;
+
+typedef struct _SerializedValue {
+  int32_t index;
+  int32_t source_type;
+  uint32_t source_length;
+} SerializedValue;
+
+typedef struct _SerializedOperation {
+  int32_t operation_type;
+  uint32_t input_count;
+  uint32_t output_count;
+} SerializedOperation;
+
+typedef struct _SerializedModel {
+  int32_t version;
+  int32_t operand_count;
+  int32_t value_count;
+  int32_t operation_count;
+  int32_t input_count;
+  int32_t output_count;
+  // SerializedOperand operands[operand_count];
+  // SerializedValue values[value_count];
+  // SerializedOperation operations[operation_count];
+  // uint32_t operand_dimensions[sum(dimension_count)]
+  // uint32_t value_data[sum(source_length+pad)/4]
+  // uint32_t operation_args[sum(input_count + output_count)]
+  // uint32_t model_inputs[input_count]
+  // uint32_t model_outputs[output_count]
+} SerializedModel;
+
+
+/**
+ * Get the physically stored size of a value.  All values are padded out
+ * to a multiple of 4 bytes to ensure the next value is 4-byte aligned.
+ */
+static uint32_t value_physical_size(uint32_t len) {
+  uint32_t phys = len;
+  if (len % 4 == 0) {
+    return len;
+  }
+  return len + 4 - (phys % 4);
```

Comment:
Unlikely that this will result in an overflow, but it could if your model has size 4G - 3 bytes or more.

Reply:
True, but I think we're a long way away from shipping a 4G model on NNAPI.